// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package playground;

import "iam.proto";

service Playground {
  rpc Ping(iam.PingRequest) returns (iam.PingReply) {}
  rpc SetIamPolicy(SetIamPolicyRequest) returns (SetIamPolicyReply) {}
  rpc GetIamPolicy(GetIamPolicyRequest) returns (GetIamPolicyReply) {}
  rpc CheckIamPolicy(CheckIamPolicyRequest) returns (CheckIamPolicyReply) {}

  rpc AddGroupMember(AddGroupMemberRequest) returns (AddGroupMemberReply) {}
  rpc DelGroupMember(DelGroupMemberRequest) returns (DelGroupMemberReply) {}
  rpc ListGroupMembers(ListGroupMembersRequest) returns (ListGroupMembersReply) {}

  rpc AddResource(AddResourceRequest) returns (AddResourceReply) {}
  rpc DelResource(DelResourceRequest) returns (DelResourceReply) {}
  rpc ListResources(ListResourcesRequest) returns (ListResourcesReply) {}

  rpc AddRole(AddRoleRequest) returns (AddRoleReply) {}
  rpc DelRole(DelRoleRequest) returns (DelRoleReply) {}
  rpc ListRoles(ListRolesRequest) returns (ListRolesReply) {}
}

message AddRoleRequest {
  string role_name = 1;
  repeated string permissions = 2;
}

message AddRoleReply {
}

message DelRoleRequest {
  string role_name = 1;
}

message DelRoleReply {
}

message ListRolesRequest {
  string prefix = 1;
}

message ListRolesReply {
  repeated string role_names = 1;
}

message ListGroupMembersRequest {
  string prefix = 1;
}

message ListGroupMembersReply {
  repeated string member_names = 1;
}

message ListResourcesRequest {
  string prefix = 1;
}

message ListResourcesReply {
  repeated string full_resource_names = 1;
}

message AddGroupMemberRequest {
  string member_type_name = 1;
  repeated string parent_type_names = 2;
}

message AddGroupMemberReply {
}

message DelGroupMemberRequest {
  string member_name = 1;
  string parent_name = 2;
  bool only_delete_relationship = 3;
}

message DelGroupMemberReply {
}

message AddResourceRequest {
  string resource_type_name = 1;
  string parent_type_name = 2;
  bool no_require_parent = 3;
}

message AddResourceReply {
}

message DelResourceRequest {
  string resource_type_name = 1;
}

message DelResourceReply {

}

message Binding {
	string role = 1;
	repeated string members = 2;
}

message Policy {
	repeated Binding bindings = 1;
	string etag = 2;
}

message CheckIamPolicyRequest {
	string resource = 1;
	string permission = 2;
	string identity = 3;
}

message CheckIamPolicyReply {
	bool result = 1;
}

message SetIamPolicyRequest {
	string resource = 1;
	Policy policy = 2;
}

message SetIamPolicyReply {
	bool accepted = 1;
}

message GetIamPolicyRequest {
	string resource = 1;
}

message GetIamPolicyReply {
	string resource = 1;
	Policy policy = 2;
}
